daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInfo.scala

```diff
@@ -34,33 +34,36 @@ private[daml] class PackageInfo(pkgSignature: Map[Ref.PackageId, Ast.GenPackage[
     * Note that while all interfaces are defined in `pkgSignature`, template may not
     * be.orm
     */
-  def interfacesRetroactiveInstances: Relation[Ref.Identifier, Ref.Identifier] =
+  def interfacesRetroactiveInstances: Relation[Ref.TypeConName, Ref.TypeConName] =
     Relation.from(
       interfaces.flatMap { case (ifaceId, iface) =>
         iface.coImplements.keysIterator.map(tmplId => ifaceId -> tmplId)
       }
     )
 
   /* Union of interfacesDirectImplementations and interfacesRetroactiveInstances */
-  def interfaceInstances: Relation[Ref.Identifier, Ref.Identifier] =
+  def interfaceInstances: Relation[Ref.TypeConName, Ref.TypeConName] =
     Relation.union(interfacesDirectImplementations, interfacesRetroactiveInstances)
 
+  def instanceInterfaces: Relation[Ref.TypeConName, Ref.TypeConName] =
+    Relation.invert(interfaceInstances)
```

Comment:
I don't like the name `instanceInterfaces`, it's not meaningful and doesn't clarify the direction of the relation. The type doesn't help, since domain and codomain have the same type. For now I'd call this `interfaceInstancesByTemplate`.

As a tangent, I wonder if what we actually want is a `Relation` that carries its inverse, then we'd just use `relation.getDomain(_)` for the usual direction and `relation.getRange(_)` or `relation.getCodomain(_)` for the backwards direction.

Reply:
I am completely fine with changing the name of `instanceInterfaces` as suggested.
In this case should we not rename `interfaceInstances` as well ? 

I was thinking also of `interfaceToInstance` and `instanceToInterface`

For implementing a Relation that carries it inverse, I am not completely sure as we will have to includes the two maps to make it efficient.

Reply:
> In this case should we not rename `interfaceInstances` as well ?

Right, we can call that one `interfaceInstancesByInterface`, but I think `interfaceInstances` is fine too.

> I was thinking also of `interfaceToInstance` and `instanceToInterface`

I don't like these since the "interface instance" is the pair of interface and template, "instance" alone doesn't make sense in this context.

> For implementing a Relation that carries it inverse, I am not completely sure as we will have to includes the two maps to make it efficient.

Yes that's true, though that comment was a tangent, not something we need to do now.

Reply:
I will apply the suggested change.